File: vol1/ch3/LeetCode1249.py
class Solution:
    def minRemoveToMakeValid(self, s: str) -> str:
        st=[]
        for i in range(0,len(s)):
            ch=s[i]
            if ch=='(':
                st.append(i)
            elif ch==')':
                if  len(st)>0 and s[st[-1]]=='(':
                    st.pop()
                else:
                    st.append(i)
        ns=list(s)
        while len(st)>0:
            i=st[-1]; st.pop()
            ns.pop(i)
        return ''.join(ns)

File: vol1/ch3/LeetCode1249.cpp
class Solution {
public:
    string minRemoveToMakeValid(string s) {
        stack<int> st;
        for (int i=0;i<s.size();i++) {
            char ch=s[i];
            if (ch=='(')
                st.push(i);
            else if (ch==')') {
                if (!st.empty() && s[st.top()]=='(')
                    st.pop();
                else
                    st.push(i);
            }
        }
        while (!st.empty()) {
            int i=st.top(); st.pop();
            s.erase(s.begin()+i);
        }
        return s;
    }
};

